ISPN-11629 Initialize DataConversion storage media type only once

danberindei opened the pull request:
https://issues.redhat.com/browse/ISPN-11629

gustavonalle commented on core/src/main/java/org/infinispan/encoding/DataConversion.java:
```diff
@@ -91,26 +94,29 @@ private DataConversion(Encoder encoder, Wrapper wrapper, boolean isKey) {
+   /**
+    * @deprecated Since 11.0, will be removed with no replacement
+    */
+   @Deprecated
```
I don't think we can deprecate this and remove with no replacement, the problem is RemoteQuery uses a complicated Wrapper that is not part of core, and it drags lots of dependencies to Hibernate Search.

gustavonalle replied:
It's yet to see how the role this wrapper will play after the Hibernate Search 6 migration though...

danberindei replied:
@gustavonalle is there a description somewhere of why we need `messageDescriptor` as a field in `ProtobufValueWrapper` instead of lazily computing it from the first few bytes of the payload when needed?

gustavonalle replied:
How could it be computed? AFAICT, ```messageDescriptor``` is the schema that describe the blob

gustavonalle replied:
@anistor can tell more about it

anistor replied:
The description is in the javadocs of ProtobufValueWrapper, I remember I wrote something there @danberindei . ProtobufValueWrapper is required for indexing. It needs that messageDescriptor to be able to decode the protobuf object. No, it cannot be really lazily computed without any external help.  ProtobufValueWrapper can obtain the type name or type id from the first few bytes but then it need a reference to a SerializationContext to be able to obtain that descriptor. So a bit more moving parts are involved here. What it the reason to remove/deprecate? And what ? :) [I did not follow discussion here and on zulip]

gustavonalle replied:
@anistor deprecation would be a method in the cache that overrides a wrapper after the cache is created, and since ```ProtobufValueWrapper``` is not part of core, it'd be more difficult to associated it with a cache at creation time rather than in the RemoteQuery server lifecycle methods, but Dan has a proposal for it

danberindei replied:
@gustavonalle I have updated the PR to just move `overrideWrapper()` from `DataConversion` to a component named `ConversionManager` for now. I got a bit carried away and made a lot more changes to the initialization of the cache wrappers, so the PR is much bigger now.

@anistor the javadoc in `ProtobufValueWrapper` says just

`The Descriptor of the message (if it's a Message and not a primitive value, or null otherwise). Transient field!`

AFAICT `WrappedMessageTagHandler` reads a `WRAPPED_DESCRIPTOR_FULL_NAME` field from the payload and computes the descriptor based on that, so I was thinking maybe `ProtobufValueWrapperFieldBridge` could also use a `WrappedMessageTagHandler` to compute a descriptor on demand.

anistor replied:
@danberindei I'm glad the very informative javadoc of ```ProtobufValueWrapper.messageDescriptor``` field piqued your interest. But you may want to read also the javadoc of the setter of this field or even follow the (unfortunately) half-broken link from that description, or alternatively read the code :).

messageDescriptor field is set by WrappedMessageTagHandler once, to (possibly) be accessed many times and avoid the penalty of being recomputed on demand many times. That field must exist to cache this expensive thing.

Two places where it is used are ProtobufValueWrapperAnalyzerDiscriminator and ProtobufValueWrapperIndexingInterceptor.

ProtobufValueWrapperFieldBridge is cache-scoped, not entry-scoped. So it must store that descriptor somewhereelse .... 

I do not understand yet what and why you need to change ProtobufValueWrapper and his many ugly friends @danberindei

danberindei replied:
@anistor I do not intend to change `ProtobufValueWrapper`, but I am curious about why it is needed, so I asked "Is there a description somewhere of why we need messageDescriptor as a field in ProtobufValueWrapper instead of lazily computing it from the first few bytes of the payload when needed?"

So far I haven't seen an explanation in the javadoc or here of *why* computing the `messageDescriptor` lazily is so expensive that it's worth keeping the extra field in the heap at all times. I was imagining that the parsing of the message descriptor would be just reading a few bytes at a known offset in the value `byte[]`, is it much more than that?

anistor replied:
The primordial need for ProtobufValueWrapper is rooted in hibernate-search 5. It needs to discriminate on the class of an object (plus eventually configured class bridges) to be able to index it. But in 6 this changes a lot. I'd keep this ProtobufValueWrapper undisturbed until we have 6 merged, then we can try to re-analyse its worthiness.

danberindei replied:
Ok, fair enough

danberindei commented:
I have rebased on #8196, restoring the removed methods from DataConversion as deprecated, and added some more commits:

ISPN-11620 Add back DataConversion methods
ISPN-11629 Deprecate DataConversion static instances
ISPN-11689 Remove AdvancedCache methods deprecated in 9.4
ISPN-11629 Deprecate AdvancedCache.withWrapping
ISPN-11629 Set component registry in EncoderCache constructor
ISPN-11629 Move conversion details out of CacheImpl and SimpleCacheImpl 

ISPN-11675 also has a separate PR: #8213

gustavonalle commented on core/src/main/java/org/infinispan/encoding/DataConversion.java:
```diff
@@ -43,152 +38,155 @@
 public final class DataConversion {
 
+   /**
+    * @deprecated Since 11.0. To be removed in 14.0, with no replacement.
```
14.0? Is this 2 years from now?  :)

danberindei replied:
I hope it's going to be sooner than 2 years, but it's the example Tristan gave in the removal policy discussion.

gustavonalle commented on core/src/main/java/org/infinispan/encoding/impl/ConversionManager.java:
```diff
@@ -0,0 +1,107 @@
+ * @since 11
+ */
+@Scope(Scopes.NAMED_CACHE)
+public class ConversionManager {
```
Not sure about this name, this class is not doing conversion, but providing info about the storage type which does not change for all decorations a cache may have at runtime.

danberindei replied:
I wasn't sure about the name either, WDYT about `StorageManager` or `StorageHelper` @gustavonalle ?

gustavonalle replied:
```StorageConfigurationManager``` ?

danberindei replied:
Sounds good, updated @gustavonalle

gustavonalle commented on core/src/main/java/org/infinispan/encoding/DataConversion.java:
```diff
@@ -43,152 +38,155 @@
-   }
-
+   @Deprecated
    public boolean isConversionSupported(MediaType mediaType) {
```
I just updated https://github.com/infinispan/infinispan/pull/8196

gustavonalle commented on core/src/main/java/org/infinispan/encoding/DataConversion.java:
```diff
@@ -43,152 +38,155 @@
+   /**
+    * @deprecated Since 11.0. To be removed in 14.0, with no replacement.
+    */
+   @Deprecated
```
I just updated #8196, could you handle it first, since you are including that commit in this PR?

gustavonalle commented:
@danberindei could you rebase please?

danberindei commented:
@gustavonalle rebased, finally. There are lots of test failures, but they seem unrelated to me.

gustavonalle commented:
merged, thanks @danberindei
